Zend/zend_API.c

```diff
@@ -501,22 +501,44 @@ ZEND_API bool ZEND_FASTCALL zend_parse_arg_long_weak(zval *arg, zend_long *dest,
 		if (UNEXPECTED(!ZEND_DOUBLE_FITS_LONG(Z_DVAL_P(arg)))) {
 			return 0;
 		} else {
-			*dest = zend_dval_to_lval(Z_DVAL_P(arg));
+			zend_long lval = zend_dval_to_lval(Z_DVAL_P(arg));
+			/* Manually check arg_num is not (uint32_t)-1, as otherwise its called by
+			 * zend_verify_weak_scalar_type_hint_no_sideeffect() */
+			if (UNEXPECTED(!zend_is_long_compatible(Z_DVAL_P(arg), lval) && arg_num != (uint32_t)-1)) {
+				zend_error(E_DEPRECATED, "Implicit conversion to int from non-compatible float %f", Z_DVAL_P(arg));
```

Comment:
This shouldn't be using `%f`, which uses limited precision. I think it's particularly important to do the print with maximum precision for these messages, because it may be exactly a difference at `10^(-16)` that makes the number not precisely representable.

I'd suggest doing the print with `%.*H` and `precision = -1`. You may want to exact functions for these two error messages to make sure they always use the same format.

It might also make sense to change the order a bit, `Implicit conversion from non-compatible float %f to int` would read more smoothly -- but I could also see an argument for your current phrasing, because it puts the variable value at the end of the message. No strong opinion either way.

Reply:
Now that there is a common function, you should also call it :D

Reply:
How did I miss that... wonder if I should introduce a similar function for float-strings?